### DISCUSSION TITLE
Help with mocking keycloak-js

### DISCUSSION THREAD

**Discussion by madtag** [keycloak/keycloak#23576] [Q&A]
Hi all, I am trying to mock keycloak-js so I can unit test my code using it.

When I try to mock init, I am not able to resolve the promise with something like the following:
`jest.mock("keycloak-js", () => {
  return jest.fn().mockImplementation(() => { //mock constructor - working
    return {
      init: jest.fn().mockResolvedValue(true) 
    }
  })
})`

I get the following whern trying to invoke init: 
TypeError: Cannot read properties of undefined (reading 'then')

      28 |       redirectUri: _appInfo.keycloak_redirect_uri,
      29 |       enableLogging: true
    > 30 |     })
         |       ^
      31 |     .then((authenticated) => {
      32 |       if (
      33 |         authenticated === true &&

Any idea what I am doing wrong here?

P.S. Implementation is working fine. Passed QA and everything. I just need to add this unit test to satisfy company criteria

**Comment by jonkoops** [comment]
This is not really a question specific to Keycloak JS, but rather to Jest and how it works internally. I would recommend that you you check out the [help page](https://jestjs.io/help) for the Jest project, and try to find an answer there.

**Reply by madtag** [reply]
Nothing I am trying from 3 days of online research is working. I was hoping someone has had success with mocking keycloak-js

**Comment by CarlRivard** [comment]
This worked for me :

In your test file, you need to create the jest mock instance:

```
jest.mock("keycloak-js", () => {
 return jest.fn().mockImplementation(() => { 
       return { init: jest.fn(),
                    login: jest.fn(), 
                    logout: jest.fn(), 
                    authenticated: true, }; 
        });
 });
```

and then in your actual test you need to mock the init promise resolved value :

`keycloakMock.init = jest.fn().mockResolvedValue(true);`

**Reply by deetz99** [reply]
Would you mind elaborating a bit more on this line? `keycloakMock.init = jest.fn().mockResolvedValue(true);`. Im running into the same issue as above. If you happen to have a test file you dont mind sharing that would be amazing.

**Reply by CarlRivard** [reply]
Hope this helps

```
it("Your test name'", async () => {
        
        // You need to wait for the init function to be resolved, this is the reason you are 
        // getting undefined when you reach then ".then((authenticated)"
        await waitFor(() => {
            keycloak.init = jest.fn().mockResolvedValue(Promise.resolve({ authenticated: true }));
        });

        render(
           // Whatever you want to render...
        );

        // Async check that keycloak.init has in fact been called before checking your 
       // rendered elements
        await act(async () => {
            expect(keycloak.init).toHaveBeenCalled();
        });
        
        // screen.debug(undefined, 300000); //To see what is actually rendered.

        expect(screen.queryByText("yourTextElement")).toBeInTheDocument();
    });
```

PS: This will only work if you are using an instance of keycloak-js inside your code, but I am sure you are if you're trying to mock it ;)

**Reply by CarlRivard** [reply]
You can play with the value of authenticated inside the waitFor too, if you need to test some cases where the user is either authenticated or not, like this : 

```
   await waitFor(() => {
            keycloak.init = jest.fn().mockResolvedValue(Promise.resolve({ authenticated: true }));
            keycloak.authenticated = true;
        });
```

**Reply by batcer** [reply]
@CarlRivard where and how do you define `keycloakMock`?
